File: android/ac-actions/src/main/kotlin/com/microsoft/adaptivecards/actions/ActionButton.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

package com.microsoft.adaptivecards.actions

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material.icons.outlined.*
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import com.microsoft.adaptivecards.core.models.ActionStyle
import com.microsoft.adaptivecards.core.models.CardAction

/**
 * Styled action button based on ActionStyle
 */
@Composable
fun ActionButton(
    action: CardAction,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    val buttonColors = when (action.style) {
        ActionStyle.Positive -> ButtonDefaults.buttonColors(
            containerColor = Color(0xFF92C353)
        )
        ActionStyle.Destructive -> ButtonDefaults.buttonColors(
            containerColor = Color(0xFFC4314B)
        )
        else -> ButtonDefaults.buttonColors()
    }

    Button(
        onClick = onClick,
        enabled = action.isEnabled,
        colors = buttonColors,
        modifier = modifier
    ) {
        val iconUrl = action.iconUrl
        if (iconUrl != null) {
            val iconVector = resolveActionIcon(iconUrl)
            if (iconVector != null) {
                Icon(
                    imageVector = iconVector,
                    contentDescription = null,
                    modifier = Modifier.size(16.dp)
                )
            } else if (iconUrl.startsWith("http://") || iconUrl.startsWith("https://")) {
                AsyncImage(
                    model = iconUrl,
                    contentDescription = null,
                    modifier = Modifier.size(16.dp),
                    contentScale = ContentScale.Fit
                )
            }
            Spacer(modifier = Modifier.width(6.dp))
        }
        Text(action.title ?: "")
    }
}

/**
 * Resolves an iconUrl to a Material icon. Supports "icon:<FluentName>" prefix.
 */
private fun resolveActionIcon(iconUrl: String): ImageVector? {
    if (!iconUrl.startsWith("icon:")) return null
    // Strip style suffix (e.g., ",Filled", ",Regular") before lookup
    val name = iconUrl.removePrefix("icon:").split(",").firstOrNull()?.lowercase() ?: return null
    return when (name) {
        "alerturgent" -> Icons.Filled.Notifications
        "alert", "bell" -> Icons.Outlined.Notifications
        "belloff" -> Icons.Outlined.Notifications
        "calendar" -> Icons.Outlined.DateRange
        "send" -> Icons.Filled.Send
        "edit" -> Icons.Outlined.Edit
        "delete" -> Icons.Outlined.Delete
        "add" -> Icons.Filled.Add
        "search" -> Icons.Filled.Search
        "share" -> Icons.Filled.Share
        "star" -> Icons.Outlined.Star
        "starfilled" -> Icons.Filled.Star
        "heart" -> Icons.Outlined.FavoriteBorder
        "heartfilled" -> Icons.Filled.Favorite
        "bookmark" -> Icons.Outlined.BookmarkBorder
        "bookmarkfilled" -> Icons.Filled.Bookmark
        "link" -> Icons.Filled.Share
        "clock" -> Icons.Filled.DateRange
        "comment" -> Icons.Filled.Email
        "thumblike" -> Icons.Outlined.ThumbUp
        "eye" -> Icons.Filled.Visibility
        "eyeoff" -> Icons.Filled.VisibilityOff
        "checkmarkcircle" -> Icons.Outlined.CheckCircle
        "dismisscircle" -> Icons.Filled.Clear
        "info" -> Icons.Outlined.Info
        "warning" -> Icons.Outlined.Warning
        "errorcircle" -> Icons.Filled.Error
        "open" -> Icons.Filled.OpenInNew
        "copy" -> Icons.Filled.ContentCopy
        "save" -> Icons.Filled.Done
        "flag" -> Icons.Outlined.Flag
        "flagfilled" -> Icons.Filled.Flag
        "location" -> Icons.Outlined.LocationOn
        "phone", "call" -> Icons.Filled.Call
        "mail" -> Icons.Outlined.Email
        "video" -> Icons.Filled.PlayArrow
        "camera" -> Icons.Filled.CameraAlt
        "attach" -> Icons.Filled.AttachFile
        "document" -> Icons.Filled.Description
        "folder" -> Icons.Filled.Folder
        "settings" -> Icons.Outlined.Settings
        "filter" -> Icons.Filled.FilterList
        "morehorizontal" -> Icons.Filled.MoreHoriz
        "chevronright" -> Icons.Filled.ChevronRight
        "chevrondown" -> Icons.Filled.ExpandMore
        "chevronup" -> Icons.Filled.ExpandLess
        "navigation" -> Icons.Filled.Navigation
        "receipt" -> Icons.Filled.Receipt
        "cart", "cartfilled" -> Icons.Filled.ShoppingCart
        "arrowreset" -> Icons.Filled.Refresh
        "toggleleft" -> Icons.Filled.ToggleOn
        else -> null
    }
}

File: ios/Sources/ACActions/ActionButton.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

import SwiftUI
#if canImport(UIKit)
import UIKit
#endif
import ACCore
import ACAccessibility
import ACFluentUI

public struct ActionButton: View {
    let action: CardAction
    let hostConfig: HostConfig
    let onTap: () -> Void

    public init(
        action: CardAction,
        hostConfig: HostConfig,
        onTap: @escaping () -> Void
    ) {
        self.action = action
        self.hostConfig = hostConfig
        self.onTap = onTap
    }

    public var body: some View {
        Button(action: onTap) {
            HStack(spacing: CGFloat(hostConfig.spacing.small)) {
                if let iconUrl = iconUrl {
                    let iconSize = CGFloat(hostConfig.actions.iconSize)
                    if iconUrl.hasPrefix("icon:") {
                        let iconName = String(iconUrl.dropFirst("icon:".count))
                        Image(systemName: Self.sfSymbol(for: iconName))
                            .resizable()
                            .aspectRatio(contentMode: .fit)
                            .frame(width: iconSize, height: iconSize)
                    } else {
                        AsyncImage(url: URL(string: iconUrl)) { image in
                            image
                                .resizable()
                                .aspectRatio(contentMode: .fit)
                                .frame(width: iconSize, height: iconSize)
                        } placeholder: {
                            EmptyView()
                        }
                    }
                }

                if let title = title {
                    Text(title)
                        .lineLimit(2)
                        .fixedSize(horizontal: false, vertical: true)
                }
            }
            .font(.system(size: 14, weight: .medium))
            .padding(.horizontal, CGFloat(hostConfig.spacing.small))
            .padding(.vertical, CGFloat(hostConfig.spacing.small) * 0.75)
            .foregroundColor(buttonForegroundColor)
            .background(buttonBackgroundColor)
            .overlay(
                RoundedRectangle(cornerRadius: CGFloat(hostConfig.cornerRadius["container"] ?? 4))
                    .stroke(buttonBorderColor, lineWidth: CGFloat(hostConfig.separator.lineThickness))
            )
            .clipShape(RoundedRectangle(cornerRadius: CGFloat(hostConfig.cornerRadius["container"] ?? 4)))
        }
        .buttonStyle(.plain)
        .disabled(!(isEnabled ?? true))
        .accessibilityAction(label: title, hint: tooltip)
    }

    private var title: String? {
        switch action {
        case .submit(let a): return a.title
        case .openUrl(let a): return a.title
        case .showCard(let a): return a.title
        case .execute(let a): return a.title
        case .toggleVisibility(let a): return a.title
        case .popover(let a): return a.title
        case .runCommands(let a): return a.title
        case .openUrlDialog(let a): return a.title
        case .resetInputs(let a): return a.title
        case .unknown: return nil
        }
    }

    private var iconUrl: String? {
        switch action {
        case .submit(let a): return a.iconUrl
        case .openUrl(let a): return a.iconUrl
        case .showCard(let a): return a.iconUrl
        case .execute(let a): return a.iconUrl
        case .toggleVisibility(let a): return a.iconUrl
        case .popover(let a): return a.iconUrl
        case .runCommands(let a): return a.iconUrl
        case .openUrlDialog(let a): return a.iconUrl
        case .resetInputs(let a): return a.iconUrl
        case .unknown: return nil
        }
    }

    private var style: ActionStyle? {
        switch action {
        case .submit(let a): return a.style
        case .openUrl(let a): return a.style
        case .showCard(let a): return a.style
        case .execute(let a): return a.style
        case .toggleVisibility(let a): return a.style
        case .popover(let a): return a.style
        case .runCommands(let a): return a.style
        case .openUrlDialog(let a): return a.style
        case .resetInputs(let a): return a.style
        case .unknown: return nil
        }
    }

    private var tooltip: String? {
        switch action {
        case .submit(let a): return a.tooltip
        case .openUrl(let a): return a.tooltip
        case .showCard(let a): return a.tooltip
        case .execute(let a): return a.tooltip
        case .toggleVisibility(let a): return a.tooltip
        case .popover(let a): return a.tooltip
        case .runCommands(let a): return a.tooltip
        case .openUrlDialog(let a): return a.tooltip
        case .resetInputs(let a): return a.tooltip
        case .unknown: return nil
        }
    }

    private var isEnabled: Bool? {
        switch action {
        case .submit(let a): return a.isEnabled
        case .openUrl(let a): return a.isEnabled
        case .showCard(let a): return a.isEnabled
        case .execute(let a): return a.isEnabled
        case .toggleVisibility(let a): return a.isEnabled
        case .popover(let a): return a.isEnabled
        case .runCommands(let a): return a.isEnabled
        case .openUrlDialog(let a): return a.isEnabled
        case .resetInputs(let a): return a.isEnabled
        case .unknown: return nil
        }
    }

    /// Button background — filled for positive/destructive, outlined for default
    private var buttonBackgroundColor: Color {
        let actionStyle = style ?? .default
        let colors = hostConfig.containerStyles.default.foregroundColors

        switch actionStyle {
        case .positive:
            return Color(hex: colors.good.`default`)
        case .destructive:
            return Color(hex: colors.attention.`default`)
        case .default, .other:
            return .clear
        }
    }

    /// Button text color — white for filled positive/destructive, accent for default
    private var buttonForegroundColor: Color {
        let actionStyle = style ?? .default
        let colors = hostConfig.containerStyles.default.foregroundColors

        switch actionStyle {
        case .positive, .destructive:
            return .white
        case .default, .other:
            return Color(hex: colors.accent.`default`)
        }
    }

    /// Button border color — matches background for filled, accent for outlined
    private var buttonBorderColor: Color {
        let actionStyle = style ?? .default
        let colors = hostConfig.containerStyles.default.foregroundColors

        switch actionStyle {
        case .positive:
            return Color(hex: colors.good.`default`)
        case .destructive:
            return Color(hex: colors.attention.`default`)
        case .default, .other:
            return Color(hex: colors.accent.`default`)
        }
    }

    /// Fluent UI icon name → SF Symbol mapping table
    private static let fluentToSFSymbol: [String: String] = [
        "Calendar": "calendar", "PeopleTeam": "person.2",
        "ArrowDown": "arrow.down", "ArrowUp": "arrow.up",
        "Link": "link", "Clock": "clock", "Send": "paperplane",
        "Edit": "pencil", "Delete": "trash", "Add": "plus",
        "Search": "magnifyingglass", "Share": "square.and.arrow.up",
        "Star": "star", "StarFilled": "star.fill",
        "Heart": "heart", "HeartFilled": "heart.fill",
        "Bookmark": "bookmark", "BookmarkFilled": "bookmark.fill",
        "Comment": "bubble.right", "ThumbLike": "hand.thumbsup",
        "Eye": "eye", "EyeOff": "eye.slash",
        "CheckmarkCircle": "checkmark.circle", "DismissCircle": "xmark.circle",
        "Info": "info.circle", "Warning": "exclamationmark.triangle",
        "ErrorCircle": "exclamationmark.circle",
        "ChevronRight": "chevron.right", "ChevronDown": "chevron.down",
        "ChevronUp": "chevron.up", "Open": "arrow.up.right.square",
        "Copy": "doc.on.doc", "Receipt": "doc.text",
        "Flag": "flag", "FlagFilled": "flag.fill",
        "Location": "location", "Phone": "phone", "Mail": "envelope",
        "Video": "video", "Camera": "camera", "Attach": "paperclip",
        "Document": "doc", "Folder": "folder", "Settings": "gearshape",
        "Filter": "line.3.horizontal.decrease", "MoreHorizontal": "ellipsis",
        "Cart": "cart", "CartFilled": "cart.fill",
        "Save": "square.and.arrow.down",
        "Navigation": "arrow.triangle.turn.up.right.diamond",
        "AlertUrgent": "bell.badge",
        "Alert": "exclamationmark.triangle",
        "Bell": "bell",
        "BellOff": "bell.slash",
        "ArrowReset": "arrow.counterclockwise",
        "ToggleLeft": "switch.2",
        "ArrowExport": "arrow.up.forward.square",
        "AccessTime": "clock",
        "Airplane": "airplane"
    ]

    /// Maps Fluent UI icon names to SF Symbols.
    /// Handles style suffixes like "Open,Filled" or "Send,Regular" by stripping the comma-separated style.
    static func sfSymbol(for fluentIcon: String) -> String {
        // Strip style suffix (e.g., ",Filled", ",Regular") if present
        let name = fluentIcon.split(separator: ",").first.map(String.init) ?? fluentIcon
        return fluentToSFSymbol[name] ?? "questionmark.square"
    }
}
